charms/istio-pilot/lib/charms/istio_pilot/v0/istio_gateway_info.py

```diff
@@ -82,44 +146,89 @@ def __init__(self, message):
 
 
 class GatewayRequirer(Object):
-    def __init__(self, charm, relation_name: str = RELATION_NAME):
-        super().__init__(charm, relation_name)
-        self.charm = charm
-        self.relation_name = relation_name
+    """Base class that represents a requirer relation end.
 
-    def get_relation_data(self):
-        if not self.model.unit.is_leader():
-            return
-        gateway = self.model.relations[self.relation_name]
-        if len(gateway) == 0:
-            raise GatewayRelationMissingError()
-        if len(gateway) > 1:
-            raise GatewayRelationTooManyError()
-
-        remote_app = [
-            app
-            for app in gateway[0].data.keys()
-            if isinstance(app, Application) and not app._is_our_app
-        ][0]
-
-        data = gateway[0].data[remote_app]
-
-        if not "gateway_name" in data:
-            logger.error(
-                "Missing gateway name in gateway-info relation data. Waiting for gateway creation in istio-pilot"
-            )
-            raise GatewayRelationDataMissingError(
-                "Missing gateway name in gateway-info relation data. Waiting for gateway creation in istio-pilot"
-            )
+    Args:
+        requirer_charm (CharmBase): the requirer application
+        relation_name (str, optional): the name of the relation
 
-        if not "gateway_namespace" in data:
-            logger.error("Missing gateway namespace in gateway-info relation data")
-            raise GatewayRelationDataMissingError(
-                "Missing gateway namespace in gateway-info relation data"
-            )
+    Attributes:
+        requirer_charm (CharmBase): variable for storing the requirer application
+        relation_name (str): variable for storing the name of the relation
+    """
+
+    def __init__(self, requirer_charm, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(requirer_charm, relation_name)
+        # TODO: remove this attribute, we are not really using it at all.
+        self.requirer_charm = requirer_charm
+        self.relation_name = relation_name
 
+    def get_relation_data(self) -> dict:
+        """Returns a dictionary with the Gateway information.
+
+        Raises:
+            GatewayRelationDataMissingError: if data is missing attributes
```

Comment:
Since this calls `relation_preflight_checks`, it might raise anything from there.  Maybe say this?


```suggestion
        Raises:
            GatewayRelationDataMissingError: if data is missing
            GatewayRelationMissingError: if there is no related application
```

Reply:
Updated